Sources/BSON/BSONDBPointer.swift

```diff
@@ -16,6 +16,71 @@ public struct BSONDBPointer: Equatable, Hashable {
 }
 
 extension BSONDBPointer: BSONValue {
+    /*
+     * Initializes a `BSONDBPointer` from ExtendedJSON.
+     *
+     * Parameters:
+     *   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for a `DBPointer`.
+     *   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+     *              This is used for error messages.
+     *
+     * Returns:
+     *   - `nil` if the provided value is not a `DBPointer`.
+     *
+     * Throws:
+     *   - `DecodingError` if `json` is a partial match or is malformed.
+     */
+    internal init?(fromExtJSON json: JSON, keyPath: [String]) throws {
+        switch json {
+        case let .object(obj):
+            // canonical and relaxed extended JSON
+            guard let value = obj["$dbPointer"] else {
+                return nil
+            }
+            guard obj.count == 1 else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Expected only \"$dbPointer\" key, found too many keys: \(obj.keys)"
+                )
+            }
+            guard let dbPointerObj = value.objectValue else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Expected \(value) to be an object"
+                )
+            }
+            guard dbPointerObj.count == 2,
+                let ref = dbPointerObj["$ref"],
+                let id = dbPointerObj["$id"] else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Expected \"$ref\" and \"$id\" keys, " +
+                        "found \(dbPointerObj.keys.count) key(s) within \"$dbPointer\": \(dbPointerObj.keys)"
+                )
+            }
+            guard let refStr = ref.stringValue,
+                let idStr = id.stringValue else {
```

Comment:
so the `$id` portion of this is actually an object id in _extended JSON_, meaning it's a `{ "$oid": <hexstring> }`, not a hex string itself. So instead of doing `stringValue` here, we can re-use the `ObjectId(fromExtJSON:)` initializer.

Reply:
okay done

Reply:
I think we can omit adding to the keyPath here actually, since we're still within the body of a single "value" type so to speak and haven't traversed any real keys. One way to think about it is that `keyPath` tracks the keys that will show up in the decoded BSON, so all the different `$<blah>` won't be there, but something like `"name"` will be.

Reply:
I would still pass in `keyPath` though right? just not add `"$dbPointer`

Reply:
Yeah, exactly.

Reply:
done